samplelistener/listener/dispatcher_service.bal

```diff
@@ -2,46 +2,43 @@ import ballerina/http;
-   //private string verificationToken;
+     private map<GenericService> services = {};
 
-   isolated function init(SlackAppMentionHandlingService|SlackAppCreatedHandlingService serviceRef) returns error? { //Receive verification token
-        self.serviceRef = serviceRef;
-        //self.verificationToken = "9asdbas9009123nas1e2";
+   isolated function addServiceRef(string serviceType, GenericService genericService) returns error? {
+        if (self.services.hasKey(serviceType)) {
+             return error("Same service type has been defined more than one time");
+        }
+        self.services[serviceType] = genericService;
    }
 
-   isolated resource function post events (http:Caller caller, http:Request request) returns error? {
+   // We are not using the (@http:payload GenericEventWrapperEvent g) notation because of 
+   // a bug in Ballerina. (Unable to handle complex objects as payloads)
+   resource function post events (http:Caller caller, http:Request request) returns error? {
         json payload = check request.getJsonPayload();
-        // string eventOrVerification = check payload.'type;
-
-        // if (payload.token !== self.verificationToken) {
-        //     return error("Verification token mismatch");
-        // }
-
-        // if (eventOrVerification == URL_VERIFICATION) {
-        //     check self.verifyURL(caller, payload);
-        // } else if (eventOrVerification == EVENT_CALLBACK) {
-        // }
 
         GenericEventWrapperEvent genericEvent = check payload.cloneWithType(GenericEventWrapperEvent);
-        if (genericEvent.event.'type == "app_mention") {
-                SlackAppMentionHandlingService serviceReference = <SlackAppMentionHandlingService> self.serviceRef;
-                var s = check self.callOnAppEvent(genericEvent, "app_mention", "onAppMention", serviceReference);
+        match genericEvent.event.'type {
+             "app_mention_added" => {
+                if self.services.hasKey("AppMentionHandlingService") {
+                    check self.callOnAppEvent(genericEvent, genericEvent.event.'type, "onAppMentionAdded", self.services["AppMentionHandlingService"]);
+               }
+             }
+             "app_mention_removed" => {
+                if self.services.hasKey("AppMentionHandlingService") {
+                    check self.callOnAppEvent(genericEvent, genericEvent.event.'type, "onAppMentionRemoved", self.services["AppMentionHandlingService"]);
+               }
+             }
+             "app_created" => {
+               if self.services.hasKey("AppCreatedHandlingService") {
+                    check self.callOnAppEvent(genericEvent, genericEvent.event.'type, "onAppCreated", self.services["AppCreatedHandlingService"]);
+               }
+             }
         }
    }
 
     isolated function callOnAppEvent(GenericEventWrapperEvent event, string eventName, string eventFunction, any serviceObj) returns error?
```

Comment:
Can't we use `GenericService` instead of `any`?
```suggestion
    isolated function callOnAppEvent(GenericEventWrapperEvent event, string eventName, string eventFunction, GenericService serviceObj) returns error?
```

Reply:
Resolved